hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AUnbuffer.java

```diff
@@ -60,6 +70,66 @@ public void testUnbuffer() throws IOException {
     }
   }
 
+  /**
+   * Test that calling {@link S3AInputStream#unbuffer()} merges a stream's
+   * {@link org.apache.hadoop.fs.s3a.S3AInstrumentation.InputStreamStatistics}
+   * into the {@link S3AFileSystem}'s {@link S3AInstrumentation} instance.
+   */
+  @Test
+  public void testUnbufferStreamStatistics() throws IOException {
+    describe("testUnbufferStreamStatistics");
+
+    // Create a new S3AFileSystem instance so we can have an independent
+    // instance of S3AInstrumentation
+    Configuration conf = createConfiguration();
+    S3AFileSystem fs = new S3AFileSystem();
+    fs.initialize(getFileSystem().getUri(), conf);
+
+    // Open file, read half the data, and then call unbuffer
+    FSDataInputStream inputStream = null;
+    try {
+      inputStream = fs.open(dest);
+
+      // Sanity check to make sure the stream statistics are 0
+      assertEquals(0,
+              fs.getInstrumentation().getCounterValue(STREAM_SEEK_BYTES_READ));
+
+      assertEquals(8, inputStream.read(new byte[8]));
+      inputStream.unbuffer();
+
+      // Validate that calling unbuffer updates the input stream statistics
+      assertEquals(8,
+              fs.getInstrumentation().getCounterValue(STREAM_SEEK_BYTES_READ));
+
+      // Validate that calling unbuffer twice in a row updates the statistics
+      // correctly
+      assertEquals(4, inputStream.read(new byte[4]));
+      inputStream.unbuffer();
+      assertEquals(12,
+              fs.getInstrumentation().getCounterValue(STREAM_SEEK_BYTES_READ));
+    } finally {
+      if (inputStream != null) {
+        inputStream.close();
+      }
+    }
+
+    // Validate that closing the file does not further change the statistics
+    assertEquals(12,
+              fs.getInstrumentation().getCounterValue(STREAM_SEEK_BYTES_READ));
+
+    // Validate that the input stream stats are correct when the file is closed
+    assertNotNull(inputStream);
+    assertEquals(12,
+            ((S3AInputStream) inputStream.getWrappedStream())
+                    .getS3AStreamStatistics().bytesRead);
+  }
+
+  @Override
+  public void teardown() throws Exception {
+    getFileSystem().delete(dest, true);
```

Comment:
superclass deleteTestDirInTeardown() should do this. If it doesn't, guard the call with checks for filesystem and destpath not being null, so it won't hide the stack traces of a failure in setup